library/Vanilla/Web/WebLinking.php

```diff
@@ -0,0 +1,103 @@
+<?php
+/**
+ * @author Alexandre (DaazKu) Chouinard <alexandre.c@vanillaforums.com>
+ * @copyright 2009-2018 Vanilla Forums Inc.
+ * @license https://opensource.org/licenses/GPL-2.0 GPL-2.0
+ */
+
+namespace Vanilla\Web;
+
+/**
+ * Class WebLinking
+ */
+class WebLinking {
+
+    /** @var array */
+    private $links = [];
+
+    /**
+     * Add a link.
+     *
+     * @link http://tools.ietf.org/html/rfc5988
+     * @link http://www.iana.org/assignments/link-relations/link-relations.xml
+     *
+     * @param string $uri Target URI for the link.
+     * @param string $rel Link relation. Either an IANA registered type, or an absolute URL.
+     * @param array $attributes Link parameters.
+     * @return WebLinking
+     */
+    public function addLink($rel, $uri, $attributes = []) {
```

Comment:
The rel should be second and optional. On the spec only the URI is required. Same for others.

Reply:
That's more or less incorrect. 

From: https://tools.ietf.org/html/rfc5988#section-3

> In this specification, a link is a typed connection between two
   resources that are identified by Internationalised Resource
   Identifiers (IRIs) [RFC3987], and is comprised of:

>   o  A context IRI,
>
>   o  a link relation type (Section 4),
>
>   o  a target IRI, and
>
>   o  optionally, target attributes.

From: https://tools.ietf.org/html/rfc5988#section-5 specifies that 

> The Link Header Field
> ... It is semantically equivalent to the <LINK> element in HTML

From: https://www.w3.org/TR/2014/REC-html5-20141028/document-metadata.html#the-link-element
> A link element must have a rel attribute.
> ...
> The types of link indicated (the relationships) are given by the value of the rel attribute, which, if present, must have a value that is a set of space-separated tokens. The allowed keywords and their meanings are defined in a later section. If the rel attribute is absent, has no keywords, or if none of the keywords used are allowed according to the definitions in this specification, then the element does not create any links.

My understanding of all this is that a link that has no rel is not a valid link.
All the examples in the spec support that conclusion and multiple implementations too.

I also put rel first because it is more convenient for the removal function.
You can remove all links of a specific relation with one call.

Reply:
While I can't find evidence in the [original Web Linking RFC](https://tools.ietf.org/html/rfc5988) explicitly stating the "relation type" is required, I also can't find a single practical example of a link header that does not include it. Also, the latest  [Web Linking RFC](https://tools.ietf.org/html/rfc8288) *does* explicitly state the following:

> The relation type of a link conveyed in the Link header field is conveyed in the "rel" parameter's value.  The rel parameter MUST be present but MUST NOT appear more than once in a given link-value; occurrences after the first MUST be ignored by parsers.

Requiring a relation type seems like a good idea to me, although I do acknowledge it might not fully adhere to the original RFC.